worker/src/main/java/org/apache/celeborn/service/deploy/worker/storage/FileWriter.java

```diff
@@ -154,19 +155,27 @@ public void decrementPendingWrites() {
 
   @GuardedBy("flushLock")
   protected void flush(boolean finalFlush) throws IOException {
+    MemCacheManager cacheManager = MemCacheManager.getMemCacheManager();
     // flushBuffer == null here means writer already closed
     if (flushBuffer != null) {
       int numBytes = flushBuffer.readableBytes();
       if (numBytes != 0) {
         notifier.checkException();
-        notifier.numPendingFlushes.incrementAndGet();
-        FlushTask task = null;
-        if (channel != null) {
-          task = new LocalFlushTask(flushBuffer, channel, notifier);
-        } else if (fileInfo.isHdfs()) {
-          task = new HdfsFlushTask(flushBuffer, fileInfo.getHdfsPath(), notifier);
+        if (finalFlush && fileInfo.getFileLength() == 0 && cacheManager.canCache(numBytes)) {
```

Comment:
Multiple thread can call `flush` simutaneously. We need synchronize `cacheManager.canCache(numBytes)`. Maybe we can add a method `cacheManager#tryCache` which returns true if caches successfully.